File: javascript/src/mediator/colleague.js
// Mediator: colleagues interact only through one object that encodes the rules.
// A mediator is any object with notify(sender, message).

/** Knows its mediator and communicates with it, never with other colleagues. */
export class Colleague {
  #mediator;
  #received = [];

  constructor(mediator) {
    this.#mediator = mediator;
  }

  send(message) {
    this.#mediator.notify(this, message);
  }

  receive(message) {
    this.#received.push(message);
  }

  received() {
    return [...this.#received];
  }

  name() {
    return this.constructor.name;
  }
}

File: python/src/patterns/mediator/colleague.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .mediator import Mediator


class Colleague:
    """Knows its mediator and communicates with it, never with other colleagues."""

    def __init__(self, mediator: Mediator) -> None:
        self._mediator = mediator
        self._received: list[str] = []

    def send(self, message: str) -> None:
        self._mediator.notify(self, message)

    def receive(self, message: str) -> None:
        self._received.append(message)

    def received(self) -> list[str]:
        return list(self._received)

    def name(self) -> str:
        return type(self).__name__
